udf/agent/server.go

```diff
@@ -32,59 +31,88 @@ func NewServer(l net.Listener, a Accepter) *Server {
 	return &Server{
 		listener: l,
 		accepter: a,
-		conns:    make(chan net.Conn),
 		stopping: make(chan struct{}),
 	}
 }
 
+// Server starts the server and blocks.
 func (s *Server) Serve() error {
+	s.mu.Lock()
+	if s.stopped {
+		s.mu.Unlock()
+		return nil
+	}
+	s.wg.Add(1)
+	s.mu.Unlock()
+
+	defer s.wg.Done()
 	return s.run()
 }
 
+// Stop closes the listener and stops all server activity.
 func (s *Server) Stop() {
 	s.mu.Lock()
-	defer s.mu.Unlock()
 	if s.stopped {
+		s.mu.Unlock()
 		return
 	}
 	s.stopped = true
+	s.listener.Close()
```

Comment:
Yes, although I don't know why :) It was being ignored previously. And is typically closed via defer, meaning its error is typically ignored.